ddclient

```diff
@@ -2307,6 +2306,25 @@ sub get_ip {
     return $ip;
 }
 
+######################################################################
+## is_ipv4() validates if string is valid IPv4 address and only
+## a valid IPv4 address (no preceding or trailing spaces/characters)
+######################################################################
+sub is_ipv4 {
+    my ($value) = @_;
+    return (length($value // '') != 0) && ($value eq ipv4_match($value));
+}
+
+######################################################################
+## ipv4_match() extracts the first valid IPv4 address from given string
+## IPv4 address may be in form nn.nn.nn.nn or nn.nn.nn.nn/nn but the
+## function only returns the IP address without the mask.
```

Comment:
Why do we need this? It makes the regex difficult to understand.

Reply:
Yes.  For example th e output from `ip -4 -o addr show dev ens160`
```
 2: ens160    inet 192.168.17.2/24 brd 192.168.17.255 scope global noprefixroute ens160\       valid_lft forever preferred_lft forever
```
And the output from `ifconfig ens160 | grep "inet "`
```
inet 192.168.17.2  netmask 255.255.255.0  broadcast 192.168.17.255
```
And I want to be able to parse out the IPv4 address from either.

Reply:
What was wrong with the original regular expression? A digit followed by `/` matches `\b`, so there should be no need to explicitly look for a CIDR mask:

```shell
$ perl -Mstrict -w -e '
  my $x = " 2: ens160    inet 192.168.17.2/24 brd 192.168.17.255 scope global noprefixroute ens160";
  if ($x =~ /\b((?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?))\b/ai) {
    printf("%s\n", $1);
  } else {
    printf("no match\n");
  }
'
192.168.17.2
```

Reply:
Here's a more compact RE:  `/\b((?:(25[0-5]|(?:2[0-4]|1\d|[1-9])?\d)\.){3}(?2))\b/a`

Reply:
So we have to decide...

1. For CIDR /xx do we care what xx is?  My regex checked that it was in the range of 0 to 32.
2. Do we allow for leading zeros within the IP... should we accept 192.168.100.001 for example or only 192.168.100.1  again my regex accepted the leading zeros.

ddclient currently has code to remove the leading zeros, but only after validating the IPv4.  And we should validate a IPv4 provided by the user where leading zeros may be provided even though not strictly legal.  So I think we need to allow for (2) but as the CIDR is irrelevant we can dump that part of the regex.

Reply:
> So we have to decide...
> 
>   1. For CIDR /xx do we care what xx is?  My regex checked that it was in the range of 0 to 32.

Nope, we don't care. It has no relevance to ddclient.

> 
>   2. Do we allow for leading zeros within the IP... should we accept 192.168.100.001 for example or only 192.168.100.1  again my regex accepted the leading zeros.

It's been ages since I've seen anyone zero-pad, and on some systems a leading 0 means octal. I guess we might as well continue to accept leading zeros given that we have historically accepted them. We can remove support for leading zeros in a follow-up commit if desired.

For the new `usev4` family of options I would prefer to reject leading zeros, if doing so wouldn't make the code overly complicated.

Reply:
I have simplified the regex.  Yes makes sense to reject leading zeros in the `usev4` version.